src/parseTools.js

```diff
@@ -1708,3 +1708,23 @@ function sendI64Argument(low, high) {
     return low + ', ' + high;
   }
 }
+
+// Add assertions to catch common errors when using the Promise object we
+// create on Module.ready() and return from MODULARIZE Module() invocations.
+function addReadyPromiseAssertions(promise) {
+  // Warn on someone doing
+  //
+  //  var instance = Module();
+  //  ...
+  //  instance._main();
+  var properties = keys(EXPORTED_FUNCTIONS);
+  // Also warn on onRuntimeInitialized which might be a common pattern with
+  // older MODULARIZE-using codebases.
+  properties.push('onRuntimeInitialized');
+  return properties.map(function(property) {
+    return "if (!Object.getOwnPropertyDescriptor(" + promise + ", '" + property + "')) {" +
```

Comment:
I think this would be more readable using a template literal:

```js
  return `if (!Object.getOwnPropertyDescriptor(${promise}, ${property})) {
  Object.defineProperty(${promise}, ${property}, { configurable: true, get: function() { abort('You are getting ${property} on the Promise object, instead of the instance. Use .then() to get called back with the instance, see the MODULARIZE docs in src/settings.js') } });" +
  Object.defineProperty(${promise}, ${property}, { configurable: true, set: function() { abort('You are setting ${property} on the Promise object, instead of the instance. Use .then() to get called back with the instance, see the MODULARIZE docs in src/settings.js') } });" +
}`;
```

Reply:
Good idea, thanks! Yeah, we should use more modern JS, now that we require Node 12.9.* at least anyhow.

Reply:
Hmm.. we still need to re-land the node version requirement: https://github.com/emscripten-core/emscripten/pull/9345/files

Reply:
Oh, I forgot about that...

@sbc100 do we need to do anything more than update all node versions in the emsdk manifest file to 12.9.1?

Reply:
No I  think we just need to update the re-land that old PR

Reply:
Ok, that sounds good in general to do.

Meanwhile I see that actually node 4 already had support for this: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Template_literals - so looks safe to use that notation here now.

Reply:
This makes me think we should have some kind of documentation on our use of JS both in the compiler output and in the compiler internals.

I think it would be very useful to be able to point to when doing JS code reviews like.   

It sounds like using template literals is fine, both in the generated code and in the compiler code?

Reply:
Not sure about the generated code. When we emit code for legacy browsers we may want to avoid that. Node 4 is 5 years old, but people may want to target even older things...

But yeah, I'll look into documenting this stuff, separately from this PR, good idea.

Reply:
IMO you shouldn't be worried about supporting platforms past their EOL if there are features that would make a real improvement to Emscripten's output. IE 11 is the only one still being supported that doesn't support template literals. If it was up to me I'd say that if anyone needs IE 11 support they can put Emscripten's output through Babel, but I'm not the one supporting Emscripten, so you choose what you want :)

Reply:
Update the range of supported targets, and possibly raising the min requirments is certainly something we can and should do.   There are users who are targeting very old and embedded environment though so expect some push back.

In any case that is a larger discussion and IMHO outside the scope of this change.

Reply:
I agree it's out of scope. I do like the idea though, and since we've improved our npm integration, I wonder if we can just emit new JS by default and run babel at the end optionally? We've discussed similar things before but it seems more practical now.